Cython/Utility/FunctionArguments.c

```diff
@@ -250,60 +278,427 @@ static int __Pyx_ParseOptionalKeywords(PyObject *kwds, PyObject *const *kwvalues
+            // No exact match found:
+            // compare against positional (always reject) and keyword (reject unknown) names.
+            #if CYTHON_AVOID_BORROWED_REFS
+            Py_INCREF(key);
+            #endif
+
+            size_t index_found = 0;
+            int cmp = __Pyx_MatchKeywordArg(key, argnames, first_kw_arg, &index_found, function_name);
+            if (cmp != 1) {
+                if (cmp == 0) {
+                    PyErr_Format(PyExc_TypeError,
+                        "%s() got an unexpected keyword argument '%U'",
+                        function_name, key);
+                }
+                #if CYTHON_AVOID_BORROWED_REFS
+                Py_DECREF(key);
+                #endif
+
+                break;
+            }
+            #if CYTHON_AVOID_BORROWED_REFS
+            Py_DECREF(key);
+            #endif
+        }
+    }
+
+    assert(PyErr_Occurred());
+}
+
+static int __Pyx_ParseKeywordDict(
+    PyObject *kwds,
+    PyObject ** const argnames[],
+    PyObject *values[],
+    Py_ssize_t num_pos_args,
+    const char* function_name,
+    int ignore_unknown_kwargs)
+{
+    PyObject** const *name;
+    PyObject** const *first_kw_arg = argnames + num_pos_args;
+    Py_ssize_t kwcount, extracted = 0;
+    int failed = 0;
+
+    // Check if dict is unicode-keys-only and let Python set the error otherwise.
+    if (unlikely(!PyArg_ValidateKeywordArguments(kwds))) return -1;
+
+    #if CYTHON_COMPILING_IN_CPYTHON && CYTHON_ASSUME_SAFE_SIZE
+    kwcount = PyDict_GET_SIZE(kwds);
+    #else
+    kwcount = PyDict_Size(kwds);
+    if (unlikely(kwcount < 0)) return -1;
+    #endif
+
+    // Extract declared keyword arguments.
+    __Pyx_BEGIN_CRITICAL_SECTION(kwds);
+    name = first_kw_arg;
+    while (*name && kwcount > extracted) {
+        PyObject * key = **name;
+        PyObject *value;
+        int found;
+
+        #if __PYX_LIMITED_VERSION_HEX >= 0x030d0000
+        found = PyDict_GetItemRef(kwds, key, &value);
+        #else
+        // There isn't a good way to make this thread-safe before Py3.13.
+        // PyObject_GetItem() would generate lots of expensive KeyErrors.
+        // PyDict_Next() returns borrowed references.
+        value = PyDict_GetItemWithError(kwds, key);
+        if (value) {
+            Py_INCREF(value);
+        } else {
+            if (unlikely(PyErr_Occurred())) {
+                failed = 1;
+                break;
+            };
+        }
+        found = value != NULL;
+        #endif
+
+        if (found) {
+            if (unlikely(found < 0)) {
+                failed = 1;
+                break;
+            };
+            values[name-argnames] = value;
+            extracted++;
+        }
+
+        #if CYTHON_COMPILING_IN_CPYTHON && CYTHON_ASSUME_SAFE_SIZE
+        kwcount = PyDict_GET_SIZE(kwds);
+        #else
+        kwcount = PyDict_Size(kwds);
+        if (unlikely(kwcount < 0)) {
+            failed = 1;
+            break;
+        };
```

Comment:
Not sure this is needed. You read it earlier and we don't really expect this to change.

If it's for thread safety reasons then I'm not sure it's a useful level of thread safety (in that there isn't really a consistent right answer for what you should get if someone's changing the dict externally as you try to extract the keywords)